storage-api/src/java/org/apache/hadoop/hive/ql/exec/vector/BytesColumnVector.java

```diff
@@ -258,73 +262,56 @@ public void setValPreallocated(int elementNum, int length) {
   public void setConcat(int elementNum, byte[] leftSourceBuf, int leftStart, int leftLen,
       byte[] rightSourceBuf, int rightStart, int rightLen) {
     int newLen = leftLen + rightLen;
-    if ((nextFree + newLen) > buffer.length) {
-      increaseBufferSpace(newLen);
-    }
-    vector[elementNum] = buffer;
-    this.start[elementNum] = nextFree;
+    ensureValPreallocated(newLen);
+    vector[elementNum] = currentValue;
+    this.start[elementNum] = currentOffset;
     this.length[elementNum] = newLen;
 
-    System.arraycopy(leftSourceBuf, leftStart, buffer, nextFree, leftLen);
-    nextFree += leftLen;
-    System.arraycopy(rightSourceBuf, rightStart, buffer, nextFree, rightLen);
-    nextFree += rightLen;
+    System.arraycopy(leftSourceBuf, leftStart, currentValue, currentOffset, leftLen);
+    System.arraycopy(rightSourceBuf, rightStart, currentValue,
+        currentOffset + leftLen, rightLen);
   }
 
   /**
-   * Increase buffer space enough to accommodate next element.
+   * Allocate/reuse enough buffer space to accommodate next element.
+   * Updates the nextFree field to point to the start of the new record.
+   * If smallBuffer is used, smallBufferNextFree is updated.
+   *
    * This uses an exponential increase mechanism to rapidly
    * increase buffer size to enough to hold all data.
    * As batches get re-loaded, buffer space allocated will quickly
    * stabilize.
    *
    * @param nextElemLength size of next element to be added
+   * @return the buffer to use for the next element
    */
-  public void increaseBufferSpace(int nextElemLength) {
-    // A call to increaseBufferSpace() or ensureValPreallocated() will ensure that buffer[] points to
+  private byte[] allocateBuffer(int nextElemLength) {
+    // A call to ensureValPreallocated() will ensure that buffer[] points to
     // a byte[] with sufficient space for the specified size.
-    // This will either point to smallBuffer, or to a newly allocated byte array for larger values.
 
-    if (nextElemLength > MAX_SIZE_FOR_SMALL_BUFFER) {
-      // Larger allocations will be special-cased and will not use the normal buffer.
-      // buffer/nextFree will be set to a newly allocated array just for the current row.
-      // The next row will require another call to increaseBufferSpace() since this new buffer should be used up.
-      byte[] newBuffer = new byte[nextElemLength];
+    // If this is a large value or small buffer is maxed out, allocate a
```

Comment:
Should we allocate another shared buffer or should we go with a single use buffer? 

If the incidence is low then it helps to stay with single use to keep the code simple or we go with the pattern that if `length > MAX_SIZE_FOR_SMALL_ITEM` then single use otherwise if `offset + length > MAX_SIZE_FOR_SHARED_BUFFER` allocate a new shared buffer.